scipy/stats/stats.py

```diff
@@ -5670,21 +5720,143 @@ def ttest_ind(a, b, axis=0, equal_var=True, nan_policy='propagate'):
+        # indices = np.argsort(random_state.random((m, n)), axis=0)
+        indices = np.array([random_state.permutation(m) for i in range(n)]).T
+    else:
+        indices = np.array(list(permutations(range(m)))).T
+    indices = indices.astype(int)
+
+    data = data.swapaxes(axis, -1)   # so we can index along a new dimension
+    data = data[..., indices]        # generate permutations
+    data = data.swapaxes(-2, axis)   # restore original axis order
+    data = np.moveaxis(data, -1, 0)  # permutations indexed along axis 0
+    return data
+
+
+def _calc_t_stat(a, b, equal_var, axis=-1):
+    """Calculate the t statistic along the given dimension"""
+    na = a.shape[axis]
+    nb = b.shape[axis]
+    avg_a = np.mean(a, axis=axis)
+    avg_b = np.mean(b, axis=axis)
+    var_a = np.var(a, axis=axis, ddof=1)
+    var_b = np.var(b, axis=axis, ddof=1)
+
+    if not equal_var:
+        denom = _unequal_var_ttest_denom(var_a, na, var_b, nb)[1]
+    else:
+        denom = _equal_var_ttest_denom(var_a, na, var_b, nb)[1]
+
+    return (avg_a-avg_b)/denom
+
+
+def _permutation_ttest(a, b, axis=0, permutations=10000, equal_var=True,
+                       random_state=None):
+    """
+    Calculates the T-test for the means of TWO INDEPENDENT samples of scores
+    using permutation methods
+
+    This test is an equivalent to `stats.ttest_ind`, except it doesn't require
+    the normality assumption since it uses a permutation test.  This function
+    is only called from ttest_ind permutations is not None.
+
+    Parameters
+    ----------
+    a, b : array_like
+        The arrays must be broadcastable, except along the dimension
+        corresponding to `axis` (the zeroth, by default).
+    axis : int, optional
+        The axis over which to operate on a and b.
+    permutations: int, optional
+        Number of permutations used to calculate p-value. If negative, use
+        all possible permutations.
+    equal_var: bool, optional
+        If False, a Welch's t-test is conducted.  Otherwise, an ordinary t-test
+        is conducted.
+    random_state : int, RandomState, or Generator, optional
+        Pseudo number generator state used for random sampling.
+
+    Returns
+    -------
+    statistic : float or array
+        The calculated t-statistic.
+    pvalue : float or array
+        The two-tailed p-value.
+
+    """
+    random_state = check_random_state(random_state)
+
+    t_stat0 = _calc_t_stat(a, b, equal_var, axis=axis)
+
+    na = a.shape[axis]
+    mat = _broadcast_concatenate((a, b), axis=axis)
+    mat = np.moveaxis(mat, axis, -1)
+    mat_perm = _data_permutations(mat, n=permutations,
+                                  random_state=random_state)
+    a = mat_perm[..., :na]
+    b = mat_perm[..., na:]
+    t_stat = _calc_t_stat(a, b, equal_var)
+
+    # Calculate the p-values
+    cmps = abs(t_stat) >= abs(t_stat0)
+    pvalues = (cmps.sum(axis=0) + 1.) / (permutations + 1.)
```

Comment:
1. Oops, this is from the old code and I didn't re-think it. It doesn't seem to make sense now, anyway. Say we asked for two permutations, and the observed t-statistic is less than or equal to one of them. I would think that indicates a p-value of 0.5, not 2/3. If you agree, please add suggestion to batch.
2. Yes. I meant to fix that. But if you like the idea [above ](https://github.com/scipy/scipy/pull/4824/files#r530052352), we can get rid of the possibility of negatives here.

Reply:
>     2. Yes. I meant to fix that. But if you like the idea [above ](https://github.com/scipy/scipy/pull/4824/files#r530052352), we can get rid of the possibility of negatives here.

I think so, I think to avoid confusion, we should refer to this as the 'exact' test. This phrase is the least ambiguous way to describe the test. If we also support the resampling/bootstrap variety here then we can refer to that as the bootstrapped version-unless you have better names. 

Don't mean to nitpick but the statistical nomenclature here is a bit confusing so I want to make this clear so we don't have (as) many questions in future. 
 
>     3. I'll double-check (so leave this unresolved) but yes, I think so because the zeroth dimension is for all the permutations. `axis` is the axis over which the t-statistic is calculated, and that is gone at this point. It gets eaten up by the mean and variance calculations.

Ok, this aligns w/what I understood too, thanks for confirming.